### PR TITLE
Update: Add type jsdoc to rule template

### PR DESCRIPTION
**Pull request by eightHundreds**

### CODE REVIEW

**Comment by bmish** [comment]
Neat! Can you explain a bit more about how this works and the benefits it provides in the PR description?

It looks like this allows code editors to see the type/structure of ESLint rule objects and thus provide information/autocomplete about the various rule fields. I like it.

Screen Shot 2021-09-18 at 12 13 18 PM

If this turns out to be the recommended way to annotate non-TypeScript rules with type information, then we could also add it to the [Working with Rules](https://eslint.org/docs/developer-guide/working-with-rules) rule example.

**Comment by eightHundreds** [comment]
This is the feature of vscode, and it seems also to be supported by webstorm. 
It can use the Typescript declaration to give JavaScript hints.

**Review by bmish — APPROVED** [review]
I like it!

File: rule/templates/_rule.js [hunk 1]
```diff
@@ -8,6 +8,9 @@
 // Rule Definition
 //------------------------------------------------------------------------------
 
+/**
+ * @type {import('eslint').Rule.RuleModule}
```
Comment (nzakas) [inline rule/templates/_rule.js:12]: Do we need to include `@types/eslint` to make this work?

Comment (bmish) [reply rule/templates/_rule.js:12]: My understanding is no. VSCode has [Automatic Type Acquistion](https://code.visualstudio.com/docs/nodejs/working-with-javascript#_typings-and-automatic-type-acquisition) where it automatically retrieves the corresponding `@types/` package for your dependencies.

Comment (nzakas) [reply rule/templates/_rule.js:12]: Huh, that’s cool. I wonder if that has any negative effect when running `tsc`? I don’t want to please VS Code only to anger TypeScript. Can you double check that?

Comment (bmish) [reply rule/templates/_rule.js:12]: These rules are JS files so I don't think anyone will be running `tsc` with them?

However, I checked what happens when using `@type` in a TS file, and TS shows a non-blocking hint:

```
JSDoc types may be moved to TypeScript types. ts(80004)
```

Comment (nzakas) [reply rule/templates/_rule.js:12]: Fair enough. I run `TSC` on JavaScript files frequently to generate .d.ts files from JSDoc comments. :)

**Pull request merged**